driver-async/src/main/com/mongodb/async/client/SingleResultCallbackSubscription.java

```diff
@@ -41,10 +41,11 @@ public void onResult(final TResult result, final Throwable t) {
                 if (t != null) {
                     onError(t);
                 } else {
+                    addToQueue(result);
                     synchronized (SingleResultCallbackSubscription.this) {
                         completed = true;
                     }
-                    addToQueue(result);
```

Comment:
Marking the observable completed, before adding the results to the queue caused the race.
If a call for more demand came in eg: `request(1)` before the results were queued, that thread would see no results and just call onComplete.

The fix was to split adding to the queue and processing the results - so now if the race happens, the thread that takes the lock when processing will see the results call onNext as required and then onComplete.